Fix return type of end()

Kijewski opened the pull request.

Kijewski committed:
Ensure for(… : zip) is even tested

simonask commented on include/rx/ranges.hpp:
```diff
@@ -219,41 +219,48 @@ constexpr auto operator|(LHS&& lhs, RHS&& rhs) noexcept {
-        return range.at_end();
+    constexpr bool operator==(const input_range_iterator &other) const noexcept {
+        RX_ASSERT(!bool(other.range));
+        return !bool(range) || range->at_end();
```
This seems like the wrong thing to do. It would often lead to `begin(range)` and `end(range)` each having their own copy of the input, which could be a `vector_range` holding on to heap memory. Or am I missing something?

simonask replied:
Ah I see you have modified the definition of `end()` to return a default-constructed range object instead. That might be OK, but still more expensive than the `input_range_iterator_end` sentinel object. Is there an issue with having heterogeneous iterator types?

Kijewski replied:
>  Is there an issue with having heterogeneous iterator types?

Yes. In C++17 `for (V v : r) …`  is syntactic sugar for

```c++
auto &&iter = r;
auto _begin = std::begin(iter), _end = std::end(iter);
for (; _begin != _end; ++_begin) {
    V v = *_begin;
    …;
}
```

That is, _begin and _end have the same type. This definition was relaxed in C++20, where the second row was replaced with `auto _begin = std::begin(iter); auto _end = std::end(iter);`.

simonask replied:
No, that requirement was [already relaxed](https://en.cppreference.com/w/cpp/language/range-for) in C++17. :-)

The only thing that C++20 adds is an optional "init" statement.

simonask replied:
It is a minor thing in my opinion, since I think non-default-constructible range types (for instance, various predicates) basically don't exist in the wild, but it may be best to not artificially introduce the requirement.

Kijewski replied:
I used `RX_OPTIONAL<R> range`, so R does not need to be default constructible. Should I use some other member name to make it clear that `range` is kinda "special"?

Kijewski committed:
Fix return type of end()

In C++17 a range-for loop `for (T v : iterable) b;` is syntactic sugar
for

```c++
auto &&_range = iterable;
auto _it = begin(_range), _end = end(_range);
for (; _it != _end; ++_it) b;
```

I.e. decltype(_it) == decltype(_end) == common type of begin and end.

This patch lets input_range_iterator store an `optional<R>`. An empty,
default constructed `optional<R>` is used for the end.

Kijewski commented:
I rebased the PR on #10, and throw out all unrelated stuff.

Kijewski commented:
Rebased on #10.

Right now I don't have a Windows installation handy. Could you please test if my PR actually fixes the problem in MSVC. :)

simonask commented on include/rx/ranges.hpp:
```diff
@@ -258,35 +258,47 @@ constexpr auto operator|(LHS&& lhs, RHS&& rhs) noexcept {
         class = std::enable_if_t<std::is_lvalue_reference_v<T>>>
     [[nodiscard]] constexpr auto operator-> () const noexcept {
-        return &range.get();
+        return &**this;
```
Might be worth adding an assertion for completeness here like you have in many other places, since `std::optional<T>::operator->` does not check presence.

Kijewski replied:
The checking is already done in operator*().

simonask commented:
I can test on MSVC, but I'm not sure whether it makes a difference. Heterogeneous iterators are generally supported in C++17, so if that's the problem for MSVC, it would definitely be a conformance problem. Not that that would be very surprising...

Just to clarify, the only problem I have with this PR is that it introduces a requirement that the range type is default-constructible. This is an issue because things like lambda function objects are not default-constructible, and this library heavily encourages lambdas.

Kijewski commented:
> Just to clarify, the only problem I have with this PR is that it introduces a requirement that the range type is default-constructible.

It should not add this requirement, since `std::optional<>` as used in `RX_OPTIONAL<R> range;` is default constructible.

Kijewski commented:
**AHHHHH!** I found the *actual* error: If you rename the field in_groups_of<>::Range::end it works. With the current implementation.

MSVC simply tries to invoke a bool. Well ... turns out a bool is not a function.

I'll open a new PR to fix that.
